pkg/occlient/occlient.go

```diff
@@ -3299,3 +3387,105 @@ func (c *Client) isResourceSupported(apiGroup, apiVersion, resourceName string)
 	}
 	return false, nil
 }
+
+// UpdateImageStreamOwnerReference
+func (c *Client) UpdateImageStream(imageStream *imagev1.ImageStream) (err error) {
+	_, err = c.imageClient.ImageStreams(c.Namespace).Update(imageStream)
+	return
+}
+
+func (c *Client) GetApplicationURLFromService(applicationName string) (fullURL string) {
+	service, err := c.GetServiceFromName(applicationName)
+	if err != nil {
+		return ""
+	}
+
+	if service.Status.LoadBalancer.Size() > 0 {
+		fullURL = fmt.Sprintf("http://%s:%d", service.Status.LoadBalancer.Ingress[0].Hostname, service.Spec.Ports[0].NodePort)
+	}
+
+	return fullURL
+}
+
+func (c *Client) GetApplicationURL(applicationName, labelSelector string) (fullURL string, err error) {
+	routeSupported, _ := c.IsRouteSupported()
+	if routeSupported {
+		routes, err := c.ListRoutes(labelSelector)
+		if err != nil || len(routes) <= 0 {
+			// No URL found - try looking in the Service
+			fullURL = c.GetApplicationURLFromService(applicationName)
+			if fullURL == "" {
+				// still no URL found - try looking for a knative Route therefore need to wait for Service and Route to be setup.
+				knGvr := schema.GroupVersionResource{Group: "serving.knative.dev", Version: "v1", Resource: "routes"}
+				knRoute, err := c.waitForManifestDeployCompletion(applicationName, knGvr, "Ready")
+				if err != nil {
+					return "", errors.Wrap(err, "error while waiting for deployment completion")
+				}
+				fullURL = knRoute.UnstructuredContent()["status"].(map[string]interface{})["url"].(string)
+			}
+		} else {
+			if len(routes) == 1 {
+				fullURL = fmt.Sprintf("%s://%s", getRouteProtocol(routes[0]), routes[0].Spec.Host)
+			} else {
+				return "", errors.New("multiple routes found")
+			}
+		}
+	} else {
+		// TODO: Look for other resources, ie Ingress
+		return "", errors.New("Route not supported")
+	}
+
+	if fullURL == "" {
+		return "", errors.New("URL not able to be found")
+	}
+
+	return fullURL, nil
+}
+
+// Create a function to wait for deploment completion of any unstructured object
+func (c *Client) waitForManifestDeployCompletion(applicationName string, gvr schema.GroupVersionResource, conditionTypeValue string) (*unstructured.Unstructured, error) {
+	klog.V(4).Infof("Waiting for %s manifest deployment completion", applicationName)
+	w, err := c.dynamicClient.Resource(gvr).Namespace(c.Namespace).Watch(metav1.ListOptions{FieldSelector: "metadata.name=" + applicationName})
```

Comment:
from what I understood - this function is only being used to check if `knative` route is created or not. Can we actually use knative's generated library here. the dynamic client's purpose was to be used at places where the resource is provided by the user and hence cannot be associated to strongly typed resource.

Reply:
Thanks for the feedback @girishramnani 
This may be better suited for #3478 as it was introduced there and just carried over into this PR